### DISCUSSION TITLE
Technical questions

### DISCUSSION THREAD

**Discussion by MichalLaczynskiAmartus** [MEF-GIT/MEF-LSO-Legato-SDK#33] [Team Posts]
Hello All!,

I have received a few questions from one on MEF members. I couldn't provide a quick response on my own. Maybe you can help. Thanks!

1. I see there is a JSONSchema for payload governance.
Why this and not a YANG definition?
How was the JSONSchema created? Generated from the YANG or hand-crafted to match?

2. I since found that there is a pyang plugin that was written that generates the jsonschema block in the swagger file.
That’s moved to be owned by ONAP – correct?
What’s the current “official” version of that pyang plugin?

3. Why use jsonschema instead of using YANG directly? 
Or are clients using both the js for client generation and then using YANG validation on the payload?
I’d be interested in what the community is doing with this.

I'd appreciate your help,
Thanks,

Michał

**Comment by jtpugac** [comment]
Michal,

Here are my response to your questions:


  1.  We should have governance for both JSON schemas and YANG models. To date the JSON schemas that I have worked on have been created by hand using the Papyrus UML as a guide.  The Papyrus UML is based on the specific MEF Specification (i.e., MEF 82 SD-WAN Information Model is based on MEF 70). Karthik and I have discussed developing a tool which would generated JSON schema from markdown version of MEF Specification.  Specifically, the tables in the MEF Specification would be generated into JSON schemas.
  2.  The ONAP tool we have used is the EAGLE tool which generates YANG from Papyrus UML.  There are some issues that do require additional work to modify the YANG. I don’t know the latest version of EAGLE tool (and corresponding pyang plugin).
  3.  To date we are only doing YANG and RESTCONF YANG APIs at Presto.  The remainder APIs are REST/JSON.

Thanks,

Jack

**Comment by MichalLaczynskiAmartus** [comment]
Thanks Jack!,

And one more response I got via email:
"The answer is that we agreed to align with ONAP external API which in turn is based on TMF APIs which have adopted swagger/OpenAPI/JSONschema",

Best Regards,
Michał

**Comment by mrehder** [comment]
thanks for the replies.

I agree the Swagger/OpenAPI "weight" must be considered and so JSONSchema compatibility is very important.
Also the amount of code-generation etc. available for JSONSchema compared to YANG.
Note that I have some suggestions on tightening up the JSONSchema definition (from what I've seen).

YANG to JSONSchema is mostly doable but there are some things which can't be translated (having built a high-quality translator).
I suggest that guidelines be issued for YANG usage to ensure untranslatable YANG isn't employed.
If multiple languages are to be used, there should be governance to ensure compatibility.
I'd be happy to help technically with this.

A good goal would be to ensure that the JSON conformant to the JSONSchema can be clearly translated into XML for typical YANG back-end validation (Relax-NG, Schematron/XSL as pyang generates). That would essentially enable stronger YANG validation as an option - "bilingual" schema validation.

**Comment by MichalLaczynskiAmartus** [comment]
Hi Mike,

Thanks for the help proposal.
I suggest joining the Legato Service API Call to talk about your suggestions.
https://wiki.mef.net/display/LSO/Legato+Service+API+Call+Notes-2020Q3
Please note that this week call is cancelled gut to [MEF Virtual Workshop ](https://wiki.mef.net/display/SQM/MEF+3.0+Virtual+Workshop+-+Tools+for+Inter-Provider+Service+Automation)

Regards,
Michał

**Comment by mrehder** [comment]
I have follow-up questions.

4. I see there is an evc_ovc json schema (swagger) file. However, I don't see an EVC there (no "evc" object). Did I miss something there? The endpoint objects say they are bilingual evc/ovc so it's a little confusing.

5. Is there a YANG model for OVC and EVC - a combined definition? If the answer is "generate it from UML", is there a way to have that YANG made available in github, as generating that from Papyrus is quite a learning curve.

6. is the original EVC YANG model now obsolete? I see subtle differences between that and the JSONSchema definitions in evc_ovc.json

**Comment by MichalLaczynskiAmartus** [comment]
What I have managed to get to know:

We have used MSM MEF W7.4 and corresponding JSON schemas for Carrier Ethernet and Carrier Ethernet SLS for Legato and it was also used to help with Sonata Access E-Line schemas.

I am not sure which evc_ovc.json schema is being referred to.

The original EVC YANG I believe is not up to date based on MEF 10.4 and MEF 26.2.  The schemas mentions above are the latest.

A decision in the past was made to shift Legato APIs/SDK to TMF/ONAP based and hence we do not plan on promoting Yang@Legato anymore.
